Suppressing hundreds of warnings at compile time.

aimichal opened the pull request:
Fixing all "Multi-line if / else statements must have braces" problems because it's easy to do, and makes the code clearer. Leaving everything else alone, and writing my reasoning in build.sbt.

tests pass and compile runs without noise:

```
> compile
[info] Updating {file:/Users/michalg/code/openie-standalone/}openie...
[info] Resolving jline#jline;2.12 ...
[info] Done updating.
[info] Compiling 115 Scala sources to /Users/michalg/code/openie-standalone/target/scala-2.11/classes...
[success] Total time: 17 s, completed Jun 8, 2016 5:41:34 PM
```

aimichal committed:
Suppressing hundreds of warnings at compile time. Fixing all "Multi-line if / else statements must have braces" problems (it's easy to do, and makes the code clearer). Leaving everything else alone, and writing my reasoning in build.sbt.

danxmoran commented on build.sbt:
```diff
@@ -68,7 +68,21 @@ libraryDependencies ++= Seq(
+//   side-effecting nullary methods are discouraged
+//   type parameter String defined in method apply shadows type String defined in object Predef. You may want to rename your type parameter, or possibly remove it.
+//
+// The code should be fixed to not have these problems. Until then, "-nowarn"
```
I'm worried that "Until then" will become never if all the warnings are suppressed.  For style issues that might be fine, but for these compiler warnings we probably want to leave them around to force the issue of resolving them.

How many warnings are printed if you leave out `-nowarn`?

jkinkead replied:
Yeah, +1 to this. We should not be hiding these (very real) warnings.

aimichal replied:
Here's what it looks like without -nowarn:

https://gist.github.com/aimichal/54d5f67c84d34551cbecae1452116139

My concern is that running the code with "sbt runMain edu.knowitall..." will produce this noise. That only happens the first time through the compile though, so maybe it's not as problematic as I expect.

danxmoran replied:
I think that's fine. It won't bother users of the packaged library, and it might be annoying enough to push whoever ends up maintaining the library towards gradually fixing the problems.

danxmoran commented on build.sbt:
```diff
@@ -80,3 +94,21 @@ fork in Test := true
+//   Line is more than 100 characters long
+//   Tab in Scala file
+//   Unusual class name; consider renaming
+//   illegal.imports.message
```
I wonder if we can be more granular with ignoring these?  A discrepancy between `equals` and `hashCode` feels like it could produce legitimate problems, so we might not want to hide that warning.  "Tab in Scala file" also seems like it'd be easy to fix with a global find-and-replace in IntelliJ.

Our SBT plugins already expose [a setting](https://github.com/allenai/sbt-plugins/blob/master/src/main/scala/org/allenai/plugins/StylePlugin.scala#L24) to globally enable / disable checking line length.  I think it'd be better to follow that pattern and explicitly pick out the style checks we want to disable rather than turn off everything by default.

aimichal replied:
I wondered that too. The `enableLineLimit` var is used excplictly by `configWithLineCheck`. I don't see code for other style checks, so it doesn't look like I can pick out the style checks a la cart -- at least not without updating StylePlugin.scala.

Yeah, we can fix all the style problems now but I wanted to avoid getting into the weeds at this moment. Here's what the warnings are about:
- the tab appears in a regex string (maybe "\t" is a good replacement)
- updating the hashCode/equals problems can change behavior which I wanted to isolate.
- changing unusual class names are possibly backwards incompatible changes.
- what's illegal about importing scala.collection.JavaConversions._?

I'll press on if you guys would like.

danxmoran replied:
Hrm I thought it might be possible to override the ScalastyleConfiguration from within this project's `build.sbt`, but it looks like you're right and you'd need to modify the style plugin to get that functionality (which could be useful, but probably isn't worth the time right now).

If it's quick to replace the tab characters with \t in the regex strings I think it'd be good to add in here, since it'll kill that warning and also help with the regex readability.

`scala.collection.JavaConversions._` contains a bunch of implicit conversions between conceptually equivalent Java and Scala collection types.  It's generally advised against because implicit conversions make the code much more difficult to reason about.  The better import is `scala.collection.JavaConverters._`, which through _different_ implicit magic adds the `.asJava` method to Scala collections and the `.asScala` method to Java collections.  Ideally we'd switch the imports and then go through and add the explicit conversion calls, but that might be more work than we want to do on this right now.  If so, we probably don't want to hide that warning, since the conversions are considered a dangerous feature.

Understood on your points about the class names and hashCode/equals, but I think we'd want to let those warnings keep printing.  hashCode/equals is a legitimate problem that somebody should eventually fix, and the weird class names could hurt code readability when you see `stringFormat` and expect it to be a value or method but instead it's a class.

jkinkead replied:
@aimichal : The _style_ warnings are totally fine to disable (and, generally, ignore). I'd leave the `styleCheck` override in place - it's fairly harmless - but have a simpler comment:

``` scala
// OpenIE does not conform to AI2 style guildelines (yet). Disable for now.
// TODO(michalg): Fix to conform.
(StylePlugin.StyleKeys.styleCheck in Compile) := {
  org.scalastyle.OutputResult(0, 0, 0, 0)
}
```

You could also open an issue to update the style.

The compiler warnings are a mixture of legitimate errors (hash code / equals being out-of-sync) and legitimate warnings that may eventually turn into errors (deprecation notices, advanced language features). These should either be fixed, or lived with. They inform our users of current and future _breakages_, not just bad style.

danxmoran commented on build.sbt:
```diff
@@ -80,3 +94,21 @@ fork in Test := true
+//   Classes must override both equals and hashCode, not just one
+//   Line is more than 100 characters long
+//   Tab in Scala file
+//   Unusual class name; consider renaming
```
Out of curiosity: what class name triggered this warning?

aimichal replied:
Here's the warning:

```
[warn] /Users/michalg/code/openie-standalone/src/main/scala/edu/knowitall/tool/chunk/Chunker.scala:96:8: Unusual class name; consider renaming
[warn] /Users/michalg/code/openie-standalone/src/main/scala/edu/knowitall/tool/postag/Postagger.scala:72:8: Unusual class name; consider renaming
[warn] /Users/michalg/code/openie-standalone/src/main/scala/edu/knowitall/tool/tokenize/Tokenizer.scala:56:8: Unusual class name; consider renaming
```

This is `class stringFormat` in all three files.

danxmoran replied:
Ah, it's probably complaining about the lowercase first letter.

jkinkead replied:
That's exactly what it's complaining about.

danxmoran commented:
Some comments about the general approach on my end.  I'm worried about globally suppressing everything, especially if the plan is to hand the library off to somebody else at AI2 who might later be tripped up by the lack of expected warnings.

jkinkead commented on build.sbt:
```diff
@@ -80,3 +94,21 @@ fork in Test := true
+// warnings which don't seem to affect the behavior but should be
+// fixed eventually:
+//
+//   Classes must override both equals and hashCode, not just one
```
This almost always produces bugs in the resultant code; it's not merely bad style. This should not be suppressed by disabling compiler warnings.

jkinkead commented:
I'm fine turning off the style plugin. Turning off the compiler warnings is really bad, and we should _not_ do that.

aimichal committed:
Remove -nowarn arg.

aimichal committed:
Fixing scalac warnings: "implicit conversion method ... should be enabled" and "reflective access of structural type member method ... should be enabled"

aimichal committed:
Replace tab character with "\t"

aimichal committed:
Fixing scalac warning "side-effecting nullary methods are discouraged"

aimichal committed:
Implement hashCode and/or equals as suggested by the style checker.

aimichal commented:
Made some easy-ish fixes. I'm a little rusty on equals/hashCode implementations, and relied on IntelliJ to generate `hashCode` where I could. In `SrlExtraction.scala` I had to do it manually.

Remaining scalac warnings (always shown because "-nowarn" was removed):

```
[warn] .../chunkedextractor/Relnoun.scala:130: possible missing interpolator: detected an interpolated expression
[warn] .../chunkedextractor/Relnoun.scala:135: possible missing interpolator: detected an interpolated expression
[warn] .../chunkedextractor/Relnoun.scala:139: possible missing interpolator: detected an interpolated expression
[warn] .../chunkedextractor/Relnoun.scala:406: possible missing interpolator: detected an interpolated expression
[warn] .../chunkedextractor/Relnoun.scala:409: possible missing interpolator: detected an interpolated expression
[warn] .../chunkedextractor/Relnoun.scala:412: possible missing interpolator: detected an interpolated expression
[warn] .../chunkedextractor/Relnoun.scala:449: possible missing interpolator: detected an interpolated expression
[warn] .../chunkedextractor/Relnoun.scala:695: possible missing interpolator: detected an interpolated expression
[warn] .../collection/immutable/Interval.scala:407: Adapting argument list by creating a 2-tuple: this may not be what you want.
[warn] .../collection/immutable/Interval.scala:415: Adapting argument list by creating a 2-tuple: this may not be what you want.
[warn] .../collection/mutable/PrefixSet.scala:23: type parameter String defined in method apply shadows type String defined in object Predef. You may want to rename your type parameter, or possibly remove it.
[warn] .../common/Analysis.scala:52: Adapting argument list by creating a 2-tuple: this may not be what you want.
[warn] .../common/Analysis.scala:86: Adapting argument list by creating a 3-tuple: this may not be what you want.
[warn] .../srlie/SrlExtraction.scala:257: a type was inferred to be `Any`; this may indicate a programming error.
[warn] .../srlie/SrlExtractor.scala:123: method deserialize in object DependencyGraph is deprecated: Use StringFormat instead.
[warn] .../srlie/SrlExtractor.scala:139: method serialize in class DependencyGraph is deprecated: Use StringFormat instead.
[warn] .../srlie/confidence/AnalyzeFeatures.scala:27: method deserialize in object DependencyGraph is deprecated: Use StringFormat instead.
```

Remaining style warnings (when style checking not faked out in build.sbt):

```
[warn] .../tool/chunk/Chunker.scala:96:8: Unusual class name; consider renaming
[warn] .../tool/postag/Postagger.scala:72:8: Unusual class name; consider renaming
[error] .../tool/tokenize/ClearTokenizer.scala:5:0: illegal.imports.message
[error] .../tool/tokenize/OpenNlpTokenizer.scala:6:0: illegal.imports.message
[error] .../tool/tokenize/PennTokenizer.scala:5:0: illegal.imports.message
[warn] .../tool/tokenize/Tokenizer.scala:56:8: Unusual class name; consider renaming
```

... and 335 occurrences of "Line is more than 100 characters long"

danxmoran commented on build.sbt:
```diff
@@ -80,3 +80,19 @@ fork in Test := true
+// result. The goal is to not spam the console when "sbt runMain ..." is used
+// on a fresh checkout.
+(StylePlugin.StyleKeys.styleCheck in Compile) := {
+  org.scalastyle.OutputResult(0, 0, 0, 0)
```
Final comment from me: if the majority of the warning noise is now from line lengths, you could just use our existing SBT setting to disable that specific check and allow the few other style warnings to spit out rather than nerfing the style check entirely.

danxmoran commented on src/main/scala/edu/knowitall/tool/parse/graph/DependencyNode.scala:
```diff
@@ -4,11 +4,10 @@ package parse
-import scala.collection.immutable.SortedSet
 import edu.knowitall.collection.immutable.Interval
-import tool.stem.{ Stemmer, IdentityStemmer }
+import tool.stem.{ Stemmer }
```
You can drop the `{` and `}` here.

danxmoran commented:
Just a few final nitpicks :) but nothing blocking!  I'll leave the magic acronym to @jkinkead.

aimichal committed:
Actually just supress the line length warnings.

aimichal committed:
Dropping curly braces.

jkinkead commented on src/main/scala/edu/knowitall/collection/immutable/graph/Graph.scala:
```diff
@@ -72,8 +72,12 @@ class Graph[T](
-      else rec(vertices ++ neighbors, neighbors -- vertices)
+      if (neighbors.isEmpty) {
+        vertices
+      }
```
In the future, feel free to cuddle the `else`; that's what we usually do, and it saves vertical space:

``` scala
if (neighbors.isEmpty) {
  vertices
} else {
  rec(vertices ++ neighbors, neighbors -- vertices)
}
```

jkinkead commented on src/main/scala/edu/knowitall/srlie/SrlExtraction.scala:
```diff
@@ -322,6 +357,8 @@ object SrlExtraction {
       Relation(this.text + " " + other.text, None, this.tokens ++ other.tokens, this.intervals ++ other.intervals)
     }
+
+    override def hashCode: Int = super.hashCode
```
This is wrong. You need to ensure that equality implies equal hash codes; that is, if two objects are equal via `equals`, they will always hash to the same location. This looks to be delegating to `Object.equals`, which compares object identity, not value equality.

jkinkead replied:
Ahh, you're delegating to the multipart hashCode.. That should be OK.

jkinkead replied:
It still worries me greatly to rely on the super implementation not changing.

jkinkead commented on src/main/scala/edu/knowitall/srlie/SrlExtraction.scala:
```diff
@@ -196,6 +216,16 @@ object SrlExtraction {
     def tokenSpan = span
 
     override def hashCode = intervals.hashCode * 39 + tokens.hashCode
+    def canEqual(that: Any): Boolean = that.isInstanceOf[Context]
```
`Context`? Not `MultiPart`?

aimichal replied:
Copied it from Context, and updated the equals implementation but not canEqual. I'll fix.

jkinkead commented on src/main/scala/edu/knowitall/srlie/SrlExtraction.scala:
```diff
@@ -222,6 +252,11 @@ object SrlExtraction {
     }
+
+    override def hashCode(): Int = {
+      val state = Seq(super.hashCode(), text, tokens, intervals)
```
I wouldn't bother including the super hash code, since you're already using all of the equality fields.

aimichal replied:
Yep, this was auto-generated. Simplifying.

jkinkead commented on src/main/scala/edu/knowitall/tool/tokenize/PennTokenizer.scala:
```diff
@@ -20,7 +20,7 @@ object PennTokenizer extends Tokenizer {
     ("""\.\.\.""", " ... "),
     ("[,;:@#$%&]", " $0 "),
-    ("""([^.]\)\([.])([])}>"']*)[ 	]*$""", "$1 $2$3 "),
+    ("""([^.]\)\([.])([])}>"']*)[ \t]*$""", "$1 $2$3 "),
```
This is a lot easier to read. :)

jkinkead commented:
A few comments. I think the `canEqual` check is a bug.

aimichal committed:
Using EqualsVerifier.

aimichal commented:
EqualsVerifier is an opinionated checker for `.equals` and `.hashCode` implementations. See here: http://www.jqno.nl/equalsverifier/

I used it to test the various properties of `.equals`: from Effective Java chapter 3, they are reflexivity, symmetry, transitivity, consistency, and non-nullity. Some checks revealed long-standing bugs. Some couldn't be honored, and resulted in `.warnings` exclusions. Here's a summary of changes:

File Interval.scala:

class Interval: This is a sealed class, and most descendants are also sealed, so `.equals` and `.hashCode` were made final. (The unsealed descendants are SingletonImpl, Empty, and Closed. I'm betting no one extended those.)

File DirectEdge.scala:

class DirectEdge has no tests because:
- UpEdge and DownEdge tests already exercise DirectEdge's `.equals`.
- As an abstract class, it's hard to test (see http://www.jqno.nl/equalsverifier/2010/11/14/abstract-delegation/). I tried several variations each of which produce different equally cryptic compiler errors:

```
  import edu.knowitall.collection.immutable.graph.Graph.Edge
  ...
  test("DirectedEdge") {
    val e1 = new Edge("foo", "bar", "fnord")
    EqualsVerifier.forClass(classOf[DirectedEdge[String]])
        .suppress(Warning.NULL_FIELDS)
        // .withPrefabValues(classOf[Edge[_]], new UpEdge(e1), new DownEdge(e1))
        // .withPrefabValues(classOf[Edge[String]], new UpEdge(e1), new DownEdge(e1))
        // .withPrefabValues(classOf[UpEdge[_]], new UpEdge(e1), new DownEdge(e1))
        // .withPrefabValues(classOf[UpEdge[String]], new UpEdge(e1), new DownEdge(e1))
        .verify
  }
```

Also tried this, which compiles, but still throws the abstract delegation exception deadend:

```
  test("DirectedEdge") {
    val e1 : Edge[String] = new Edge[String]("source", "dest", "label1")
    val e2 : Edge[String] = new Edge[String]("source", "dest", "label2")
    val ue1 = new UpEdge[String](e1)
    val ue2 = new UpEdge[String](e2)
    val de1 = new DownEdge[String](e1)
    val de2 = new DownEdge[String](e2)
    EqualsVerifier.forClass(classOf[DirectedEdge[_]])
        .suppress(Warning.NULL_FIELDS)
        .withPrefabValues(classOf[UpEdge[_]], ue1, ue2)
        .withPrefabValues(classOf[DownEdge[_]], de1, de2)
        .suppress(Warning.STRICT_INHERITANCE)
        .verify
  }
```

File SrlExtraction.scala

class MultiPart has its `.equals` and `.hashCode` fully tested by way of tests for Relation and Context.

File Graph.scala

class Graph: Fixed long-standing bug where `equals` was not checking internal fields `outgoing` and `incoming`.

File SrlExtraction.scala

class MultiPart: Used the correct class name in canEqual
class Context: Made `hashCode` use field values directly instead of relying on `super.hashCode`
class Relation: fixed long-standing bug where `equals` did not use field `sense`.
class Relation: defined a new `hashCode` in terms of class fields, not `super.hashCode`

jkinkead commented on build.sbt:
```diff
@@ -65,7 +65,8 @@ libraryDependencies ++= Seq(
   "org.slf4j" % "slf4j-api" % "1.7.5",
-  "org.specs2" % "specs2_2.11" % "2.3.13" % "test"
+  "org.specs2" % "specs2_2.11" % "2.3.13" % "test",
+  "nl.jqno.equalsverifier" % "equalsverifier" % "2.1"
```
Should probably be sorted. Also should be constrained to test-only (`% "test"` at the end).

jkinkead commented:
LGTM! Merging.
